// @digitalnative/lumen-targets/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sources = exports.Sources = exports.table = void 0;
/// Table for locating asset_id in STND with the ticker
/// Initial dataset are from https://wsb.gold/ top 10 ticker mentions list
/// Prices are represented in 15 decimals (e.g. one = 1e15)
exports.table = {
    0: "STND",
    1: "MTR",
    2: "AMC",
    3: "APHA",
    4: "TLRY",
    5: "SPY",
    6: "SNDL",
    7: "BB",
    8: "TWTR",
    9: "PLTR",
};
var Sources;
(function (Sources) {
    Sources[Sources["UNISWAP"] = 0] = "UNISWAP";
    Sources[Sources["PANCAKESWAP"] = 1] = "PANCAKESWAP";
    Sources[Sources["BINANCE"] = 2] = "BINANCE";
    Sources[Sources["HUOBI"] = 3] = "HUOBI";
    Sources[Sources["FINNHUB"] = 4] = "FINNHUB";
    Sources[Sources["NOMICS"] = 5] = "NOMICS";
    Sources[Sources["MISC"] = 6] = "MISC";
})(Sources = exports.Sources || (exports.Sources = {}));
exports.sources = {
    0: Sources.NOMICS,
    1: Sources.MISC,
    2: Sources.FINNHUB,
    3: Sources.FINNHUB,
    4: Sources.FINNHUB,
    5: Sources.FINNHUB,
    6: Sources.FINNHUB,
    7: Sources.FINNHUB,
    8: Sources.FINNHUB,
    9: Sources.FINNHUB,
};
